### DISCUSSION TITLE
Welcome to php-garmin-connect-api Discussions!

### DISCUSSION THREAD

**Discussion by stoufa06** [stoufa06/php-garmin-connect-api#5] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ajithv11** [comment]
I have successfully connected with garmin and getActivitySummary using $tokenCredentials as given in the README

But is there any way to getActivitySummary later by using connected userId , userIdentifier and Secret ?

**Reply by stoufa06** [reply]
@ajithv11 You need to store $tokenCredentials and  userId in your database and retrieve it anytime to request your activity summary

**Reply by stoufa06** [reply]
@ajithv11 Added [example](https://github.com/stoufa06/php-garmin-connect-api/tree/1.8.1/examples) in examples folder. Please checkout the new release

**Comment by shuny5** [comment]
The included example file for step #1 of OAuth 1.0a for says to use a function called getTemporaryCredentials (line 93 in example.php) which is not a function in the library source code that I can download? 

How do I get the temporary oauth_token & oath_token_secret to proceed to step #2? 
I am familiar with how OAuth 1.0a works but having a hard time implementing libraries to accomplish this task.

**Reply by stoufa06** [reply]
@shuny5  
[getTemporaryCredentials](https://github.com/thephpleague/oauth1-client/blob/v1.8.0/src/Server/Server.php#L74) function part of [`league/oauth1-client`](https://github.com/thephpleague/oauth1-client)library which is a dependency of the current library (see[ composer.json](https://github.com/stoufa06/php-garmin-connect-api/blob/master/composer.json#L8)) 

To run the [example](https://github.com/stoufa06/php-garmin-connect-api/blob/master/examples/example.php) in [examples folder](https://github.com/stoufa06/php-garmin-connect-api/tree/master/examples) you will need:

- You will need php 7.1 or more
- Clone the repository : `git clone git@github.com:stoufa06/php-garmin-connect-api.git`
- Install dependencies : `composer install`
- Create `.env` file in examples folder and set the variables : 
```
GARMIN_KEY=app-garmin-key
GARMIN_SECRET=app-garmin-secret-key
GARMIN_CALLBACK_URI=http://localhost:8080/example.php
```
- Go to examples folder and run the script:
```bash
cd path-to-cloned-library/examples
php -S localhost:8080
```
- Open [http://localhost:8080/example.php](http://localhost:8080/example.php) in browser.
<img width="1028" alt="image" src="https://user-images.githubusercontent.com/3877384/128849979-1984bbd4-5a40-4ccc-9ae2-ba0adc49a54b.png">

- Connect with garmin account

<img width="983" alt="image" src="https://user-images.githubusercontent.com/3877384/128851466-7e8139ff-941b-4655-9f8c-5a048ce372f2.png">
- At the end you will find something like this
- 
<img width="1379" alt="image" src="https://user-images.githubusercontent.com/3877384/128852339-de7d3a7f-7ca2-4057-85f0-e6cc6a0fd942.png">

**Reply by shuny5** [reply]
Thank you for such a quick response! 
I was still not able to implement it correctly so I know I am still doing something wrong.


This is the error I am getting:

(This is still occurring regardless of me not figuring out how the dotenv works properly)
Call to undefined function GuzzleHttp\Psr7\uri_for()

**Reply by stoufa06** [reply]
@shuny5 Not sure what you are trying to do here. I suggest you to try to make the example working first. Don't try to re-implement the library, there is nothing to gain from reinventing the wheel.

**Reply by stoufa06** [reply]
@shuny5 This is an easy docker example https://github.com/stoufa06/php-garmin-connect-api/discussions/11 to reproduce

**Reply by shuny5** [reply]
@stoufa06 I apologize for taking such a long time to reproduce the easy docker example but I did get it to work finally. I really appreciate the help!

**Reply by fulmar2** [reply]
shunyS - I know it has been a while... but do you remember how you got this fixed? I am having the exact same issue.

**Comment by ajithv11** [comment]
**Garmin user id  Issue**

I have used the library function to get Garmin User Profile ID  using the below code 
$userId = $server->getUserUid($tokenCredentials);  // Returns  e6989c1a-6c40-4d75-841c-9e7818ff7562   // this seems to be incorrect

When I logged in to my garmin account, the View Profile link  was found to be https://connect.garmin.com/modern/profile/3210730c-8149-46f2-89b2-64709fc256b5

I am not sure why the above function returns **e6989c1a-6c40-4d75-841c-9e7818ff7562**    Instead of **3210730c-8149-46f2-89b2-64709fc256b5**   

Please let me know If there is some issue or If I have missed something


Thanks in advance
Ajith

**Reply by stoufa06** [reply]
@ajithv11 I think you need to ask the garmin support. I think the apps api and website has different keys.
I have implemented it according to the docs and it is working :

> ## Get User ID
> Each Garmin Connect user has a unique API ID associated with them that will persist across multiple UATs. For instance, if a user deletes their association through Garmin Connect and then, later, completes the OAuth process to generate a new User Access Token with the same Garmin Connect account, the second token will still have the same API User ID as the first token. Similarly, if a partner is managing multiple programs and the user signs up for each of them, the API User ID returned for each of the UATs will match.
> The API ID provides no identifying information and is not used in any other Garmin API, web service, or system. There is no reason to ever pass the API User ID back to the API as user lookup will always be performed using the User Access Token in the Authorization header.
> Request URL to fetch API User ID
>
> GET https://apis.garmin.com/wellness-api/rest/user/id No parameters are required for this request.
> Response: {"userId": "d3315b1072421d0dd7c8f6b8e1de4df8"}

**Comment by ccalisi** [comment]
Can anyone help me.  I have two issues.  ( I am on PHP V8.x I believe)

1. I get this warning filling my error log. Would love to know how to fix it.

PHP Warning:  League\OAuth1\Client\Signature\HmacSha1Signature::League\OAuth1\Client\Signature\{closure}(): Argument #2 ($value) must be passed by reference, value given in /home/hzp9n88nz96p/vendor/league/oauth1-client/src/Signature/EncodesQuery.php on line 33

2. Most importantly, I cant get data.  When I try to pull data I get this:

Received error [{"errorMessage":"[1cf00fd8-6c33-4b7f-8e81-4991aa8473bb]duplicate backfill processed at 2022-06-13T16:20:30Z [2022-06-12T16:20:30Z to 2022-06-13T16:20:30Z]"}] with status code [409] when requesting historic activities summary. in /home/hzp9n88nz96p/vendor/stoufa06/php-garmin-connect-api/src/GarminApi.php:289

here is my code ( it is exactly from the example)

   $uploadStartTimeInSeconds = (new DateTime())->modify('-1 day')->getTimestamp(); // start time in seconds 
    $uploadEndTimeInSeconds = (new DateTime())->getTimestamp(); // end time in seconds

    // Backfill activities before pulling activities (probably you must wait before it fills the summaries)
    $params = [
        'summaryStartTimeInSeconds' => $uploadStartTimeInSeconds, // time in seconds utc
        'summaryEndTimeInSeconds' => $uploadEndTimeInSeconds // time in seconds utc
    ];
    $server->backfillActivitySummary($tokenCredentials, $params);

    // User id
    $userId = $server->getUserUid($tokenCredentials);
    
    // Activity summaries
    $params = [
        'uploadStartTimeInSeconds' => $uploadStartTimeInSeconds, // time in seconds utc
        'uploadEndTimeInSeconds' => $uploadEndTimeInSeconds // time in seconds utc
    ];
    $summary = $server->getActivitySummary($tokenCredentials, $params);

**Reply by stoufa06** [reply]
Update to latest version and check examples with docker : https://github.com/stoufa06/php-garmin-connect-api/discussions/20
